subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java

```diff
@@ -143,7 +146,11 @@ public void compile(List<String> files, CompilationUnit cu) {
                         if (shouldProcessAnnotations) {
                             // In order for the Groovy stubs to have annotation processors invoked against them, they must be compiled as source.
                             // Classes compiled as a result of being on the -sourcepath do not have the annotation processor run against them
-                            spec.setSource(spec.getSource().plus(new SimpleFileCollection(stubDir).getAsFileTree()));
+                            Set<File> newSource = ImmutableSet.<File>builder()
+                                .addAll(spec.getSource())
+                                .add(stubDir)
+                                .build();
+                            spec.setSource(ImmutableFileCollection.of(newSource).getAsFileTree().getFiles());
```

Comment:
This still looks weird to me. This looks like a sign that `FileOnlyImmutableFileCollection` should override `getAsFileTree()` perhaps? Let's discuss.

Reply:
after further discussion, we've decided to live with the current code.  as part of future cleanup in the Scala, Java & Groovy compiler classes, the handling of the `source` file collection & classpath should be revisited to remove the roundabout way they are configured.